copycat/api/src/main/java/org/apache/kafka/copycat/data/SchemaProjector.java

```diff
@@ -0,0 +1,209 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements.  See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.  You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ **/
+
+package org.apache.kafka.copycat.data;
+
+import org.apache.kafka.copycat.data.Schema.Type;
+import org.apache.kafka.copycat.errors.SchemaProjectorException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * <p>
+ *     SchemaProjector is utility to project a value between compatible schemas and throw exceptions
+ *     when non compatible schemas are provided.
+ * </p>
+ */
+
+public class SchemaProjector {
+
+    /**
+     * @param writer the schema that is used to write the record
+     * @param reader the schema that is used to read the record
+     * @param record the value to project from writer schema to reader schema
+     * @return the projected value with reader schema
+     * @throws SchemaProjectorException
+     */
+    public static Object project(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.isOptional() && !reader.isOptional()) {
+            if (reader.defaultValue() != null) {
+                if (record != null) {
+                    return projectRequiredSchema(writer, reader, record);
+                } else {
+                    return reader.defaultValue();
+                }
+            } else {
+                throw new SchemaProjectorException("Writer schema is optional, however, reader schema does not provide a default value.");
+            }
+        } else {
+            if (record != null) {
+                return projectRequiredSchema(writer, reader, record);
+            }
+            return null;
+        }
+    }
+
+    private static Object projectRequiredSchema(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        Object result = null;
+        // handle logical types
+        if (reader.name() != null) {
+            return projectLogical(writer, reader, record);
```

Comment:
Do we even need special handling for logical types? Shouldn't they just be like any other type where we check compatibility and otherwise the data just passes through?